### PR TITLE
BREAKING CHANGE: Add function key to payload for function calling

### PR DESCRIPTION
**Pull request by michaelpoluektov**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

Discussions entry: https://github.com/open-webui/open-webui/discussions/3586

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [ ] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [ ] **Testing:** Have you written and run sufficient tests for validating the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To cleary categorize this pull request, prefix the pull request title, using one of the following:
- **BREAKING CHANGE:** Significant changes that may affect compatibility
- **build:** Changes that affect the build system or external dependencies
- **ci:** Changes to our continuous integration processes or workflows
- **chore:** Refactor, cleanup, or other non-functional code changes
- **docs:** Documentation update or addition
- **feat:** Introduces a new feature or enhancement to the codebase
- **fix:** Bug fix or error correction
- **i18n:** Internationalization or localization changes
- **perf:** Performance improvement
- **refactor:** Code restructuring for better maintainability, readability, or scalability
- **style:** Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)
- **test:** Adding missing tests or correcting existing tests
- **WIP:** Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Breaking Change

- The `body` parameter passed to functions now contains `title`, `task` and `function` fields, which may break existing functions passing `body` directly to OpenAI API calls.

### Description

- Allow functions and pipelines to tell when they're being used for tool calling

### Added

- Add "function" key to payload for function calling

### CODE REVIEW

**Comment by michaelpoluektov** [comment]
This now breaks existing function, including default OpenAI, but makes functions more compatible with pipelines.

File: backend/main.py [hunk 1]
```diff
@@ -832,12 +833,13 @@ def filter_pipeline(payload, user):
...
                 pass
 
-    if "pipeline" not in app.state.MODELS[model_id]:
-        if "title" in payload:
-            del payload["title"]
-
-        if "task" in payload:
-            del payload["task"]
+    keep_extras = (
+        "pipeline" in app.state.MODELS[model_id] or "pipe" in app.state.MODELS[model_id]
```
Comment (tjbck) [inline backend/main.py:837]: Only pipelines should have additional fields and `pipe` function should not have any extras.

**Comment by tjbck** [comment]
The rest LGTM!

File: backend/main.py (hunk 1 above)
Comment (michaelpoluektov) [reply backend/main.py:837]: Is there any other way to know whether a function is being called for title/task/tool use? Currently I'm just checking for keywords in the prompt but I feel like that's not particularly robust.

Comment (tjbck) [reply backend/main.py:837]: I'll try to think of something for this

Comment (michaelpoluektov) [reply backend/main.py]: OK thanks! I've reverted this for now.

**Comment by tjbck** [comment]
Closing in favour of https://github.com/open-webui/open-webui/commit/c83704d6ca6584448209e9a787754239dc73641d, Added you as a co-author!

**Pull request closed without merging**